GenericList/string_list.cpp

```diff
@@ -4,47 +4,96 @@
 #include <iostream>
 #include <stdexcept>
 #include <string>
+#include <memory>
+#include <cassert>
 
 using namespace std;
 
-#include "string_list.h"
+#include "string_list.h"'
+
+std::allocator<std::string> List_String::dataAllocator;
 
 List_String::List_String() : data(nullptr), capacity(0), count(0) {
 
 }
 
+
+void swap(List_String& first, List_String& second) {
+	//explanation: calling swap makes it unqualified, meaning if swap(T,T) is specially defined, it is called, otherwise, call std::swap(T,T)
+	swap(first.capacity, second.capacity);
+	swap(first.count, second.count);
+	swap(first.data, second.data);
+}
+//Rule of 3
+
+//Destructor
 List_String::~List_String() {
-	delete[] data;
+	for (int i = 0; i < count; i++) {
+		data[i].~string();
+	}
+	dataAllocator.deallocate(data, capacity);
 }
 
-int List_String::Capacity() { return capacity; }
+//Copy constructor
+List_String::List_String(const List_String& other) : capacity(other.capacity), count(other.count) {
+	data = CreateDeepCopy(other.data, other.capacity, other.capacity);
+}
 
-bool List_String::Capacity(int new_capacity) {
-	if (new_capacity < 0) return false; //allowed values must be >= 0
-	if (new_capacity < count) return false; //can't have shorter capacity than element count
-	if (new_capacity == capacity) return true; //no change
-	if (new_capacity == 0) { //empty non-empty array
-		delete data;
-		return true;
-	};
+//Copy assignement
+List_String& List_String::operator=(const List_String& other) {
+	//Note: other is not a const reference but a copy of the value, allowing for use of swap logic (since "other" will be destroyed after
+	//Long definition:
+		//List_String tmp(other);
+		//swap(*this, tmp);
+		//return *this;
+	//clearer definition:
+	return *this = List_String(other); //calls operator=(&& List_String(&)))
 	
-	// general case: dynamic resizing
-	std::string* new_data = new std::string[new_capacity]; 
-	for (int i = 0; i < capacity; i++) {
-		new_data[i] = data[i];
+}
+
+
+//Rule of 5
+//R-value references (&&) explained: http://thbecker.net/articles/rvalue_references/section_01.html 
+//basic explanation: if a function argument is &&, whatever it references will stop existing at the end of the function
+List_String::List_String(List_String&& other) : data(nullptr), capacity(0), count(0) {
+	swap(*this, other);
+}
```

Comment:
Actually, I didn't realize that was what's happening 😅. So what you did wasn't actually wrong, I just wasn't used to seeing it. It's just my code style preference, there's probably some teams out there that do it this way, but I typically just make it really straight forward and initialize/assign the members and then clear the data on `other`. 

Also, `swap` typically requires temporary variables to do the exchange, they're cheap in this case but _may_ not be in other cases. Suppose we have something like:
```cplusplus
struct foo
{
    SomeBigClass big;
    struct bar* ptr;
    foo(foo&& other) : str(std::move(other.big)), ptr(other.ptr) {
       other.ptr = nullptr;
    }
};
```
Its more expensive to create the temporary big class within a swap. Note that I have a pointer member there, if I didn't it would be a lot simpler:
```cplusplus
struct foo
{
    SomeBigClass big;
    foo(foo&&) = default; // this is optional, if I didn't have this it would be implicitly defined
};
```